app/scripts/controllers/encryption-public-key.ts

```diff
@@ -0,0 +1,443 @@
+
+  private _getState: () => any;
+
+  private _encryptionPublicKeyManager: EncryptionPublicKeyManager;
+
+  private _messageManagers: AbstractMessageManager<
+    AbstractMessage,
+    AbstractMessageParams,
+    AbstractMessageParamsMetamask
+  >[];
+
+  private _metricsEvent: (payload: any, options?: any) => void;
+
+  /**
+   * Construct a EncryptionPublicKey controller.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the EncryptionPublicKey controller.
+   * @param options.keyringController - An instance of a keyring controller used to extract the encryption public key.
+   * @param options.getState - Callback to retrieve all user state.
+   * @param options.metricsEvent - A function for emitting a metric event.
+   */
+  constructor({
+    messenger,
+    keyringController,
+    getState,
+    metricsEvent,
+  }: EncryptionPublicKeyControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+
+    this._keyringController = keyringController;
+    this._getState = getState;
+    this._metricsEvent = metricsEvent;
+
+    this.hub = new EventEmitter();
+    this._encryptionPublicKeyManager = new EncryptionPublicKeyManager();
+
+    this._messageManagers = [this._encryptionPublicKeyManager];
+
+    const methodNames = [methodNameGetEncryptionPublicKey];
+
+    this._messageManagers.forEach((messageManager, index) => {
+      this._bubbleEvents(messageManager);
+
+      messageManager.hub.on(
+        'unapprovedMessage',
+        (msgParams: AbstractMessageParamsMetamask) => {
+          this._requestApproval(msgParams, methodNames[index]);
+        },
+      );
+    });
+
+    this._subscribeToMessageState(
+      this._encryptionPublicKeyManager,
+      (state, newMessages, messageCount) => {
+        state.unapprovedEncryptionPublicKeyMsgs = newMessages;
+        state.unapprovedEncryptionPublicKeyMsgCount = messageCount;
+      },
+    );
+  }
+
+  /**
+   * A getter for the number of 'unapproved' Messages in this.messages
+   *
+   * @returns The number of 'unapproved' Messages in this.messages
+   */
+  get unapprovedMsgCount(): number {
+    return this._encryptionPublicKeyManager.getUnapprovedMessagesCount();
+  }
+
+  /**
+   * Reset the controller state to the initial state.
+   */
+  resetState() {
+    this.update(() => getDefaultState());
+  }
+
+  /**
+   * Called when a Dapp uses the eth_getEncryptionPublicKey method, to request user approval.
+   *
+   * @param address - The address from the encryption public key will be extracted.
+   * @param [req] - The original request, containing the origin.
+   */
+  async newRequestEncryptionPublicKey(
+    address: string,
+    req: OriginalRequest,
+  ): Promise<string> {
+    const keyring = await this._keyringController.getKeyringForAccount(address);
+
+    switch (keyring.type) {
```

Comment:
Copied from metamask-controller. I don't like it either and it can be refactored before we merge.